### PR TITLE
kv: support delayed check, optimize batch insert

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: kv/union_store.go [hunk 1]
```diff
@@ -106,10 +111,39 @@ func (us *UnionStore) Delete(k []byte) error {
...
+	}
+	if len(keys) == 0 {
+		return nil
+	}
+	values, err := us.Snapshot.BatchGet(keys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for it := us.expect.NewIterator(nil); it.Valid(); it.Next() {
+		if len(it.Value()) == 0 {
```
Comment (shenli) [inline kv/union_store.go]: Not all the it.Value() in in expect is nil? 
For I see:
err = c.expect.Set(k, nil)

Comment (disksing) [inline kv/union_store.go]: Yeah, it should be nil this version. 
Here 'expect' means that we expect a particular value should be in the store, the expected value may be different in other cases later. So I checked its value for semantic integrity.

File: kv/union_store.go [hunk 2]
```diff
@@ -106,10 +111,39 @@ func (us *UnionStore) Delete(k []byte) error {
 	return us.WBuffer.Set(k, nil)
 }
 
+// CheckExpect loads all expect values from store then checks if all values are matched.
+func (us *UnionStore) CheckExpect() error {
```
Comment (c4pt0r) [inline kv/union_store.go]: LazyCheck maybe a better name?

File: kv/union_store.go [hunk 3]
```diff
@@ -39,14 +41,17 @@ func IsErrNotFound(err error) bool {
 type UnionStore struct {
 	WBuffer  MemBuffer // updates are buffered in memory
 	Snapshot Snapshot  // for read
+	expect   MemBuffer // for delay check
```
Comment (ngaut) [inline kv/union_store.go]: s/expect/conditionPairs/

Comment (ngaut) [inline kv/union_store.go]: Maybe lazyConditionPairs is a better name.

File: kv/union_store.go (hunk 1 above)
Comment (shenli) [inline kv/union_store.go]: Got it.

**Comment by shenli** [comment]
LGTM

File: kv/kv.go [hunk 4]
```diff
@@ -111,6 +111,12 @@ const (
 	// has a limit of the option value. The feature is disabled if option value <= 0 or value type is not int.
 	// This option is particularly useful when we have to do sequential Gets, e.g. table scans.
 	RangePrefetchOnCacheMiss Option = iota + 1
+
+	// PresumeKeyNotExists directives that when dealing with a Get operation and failed to read data from cache,
```
Comment (siddontang) [inline kv/kv.go]: but failing to read

File: kv/kv.go [hunk 5]
```diff
@@ -111,6 +111,12 @@ const (
...
+	// we presume that the key does not exist in Store. The actual existences will be checked before the
```
Comment (siddontang) [inline kv/kv.go]: existence

**Comment by c4pt0r** [comment]
LGTM

**Pull request merged**